naga/src/front/spv/mod.rs

```diff
@@ -560,25 +561,40 @@ struct BlockContext<'function> {
     parameter_sampling: &'function mut [image::SamplingFlags],
 }
 
+impl<'a> BlockContext<'a> {
+    /// Recurse into the expression with the given handle, locating any contained
+    /// global variables.
+    ///
+    /// This is used to track atomic upgrades.
+    fn get_contained_global_variables(
+        &self,
+        handle: Handle<crate::Expression>,
+    ) -> IndexSet<Handle<crate::GlobalVariable>> {
```

Comment:
This is more complicated than you need. You never need to look at `index` operands, so there's only ever going to be one global variable, at the most. This function can simply be a loop, not recursive, and it can simply return an `Option<Handle<GlobalVariable>>`.